lib/src/idx/trees/mtree.rs

```diff
@@ -959,57 +1305,52 @@ impl NodeVectors for InternalNode {
 		self.len()
 	}
 
-	fn get_vector(&self, i: usize) -> Result<Arc<Vector>, Error> {
-		self.get(i).ok_or(Error::Unreachable).map(|e| e.center.clone())
-	}
-
-	fn distribute_entries(
-		self,
-		distances: &[Vec<f64>],
-		p1: usize,
-		p2: usize,
-	) -> Result<(Self, f64, Self, f64), Error> {
-		let mut internal1 = InternalNode::new();
-		let mut internal2 = InternalNode::new();
-		let (mut r1, mut r2) = (0f64, 0f64);
-		for (i, r) in self.into_iter().enumerate() {
-			let dist_p1 = distances[i][p1];
-			let dist_p2 = distances[i][p2];
-			if dist_p1 <= dist_p2 {
-				internal1.push(r);
-				if dist_p1 > r1 {
-					r1 = dist_p1;
-				}
-			} else {
-				internal2.push(r);
-				if dist_p2 > r2 {
-					r2 = dist_p2;
-				}
+	fn get_objects(&self) -> Vec<Arc<Vector>> {
+		self.keys().cloned().collect()
+	}
+
+	fn extract_node(
+		&mut self,
+		distances: &DistanceCache,
+		p: Arc<Vector>,
+		a: Vec<Arc<Vector>>,
+	) -> Result<(Self, f64, Arc<Vector>), Error> {
+		let mut n = InternalNode::new();
+		let mut max_r = 0f64;
+		for o in a {
+			let mut props = self.remove(&o).ok_or(Error::Unreachable)?;
+			let dist = *distances.0.get(&(o.clone(), p.clone())).unwrap_or(&0f64);
+			let r = dist + props.radius;
+			if r > max_r {
+				max_r = r;
 			}
+			props.parent_dist = dist;
+			n.insert(o, props);
 		}
-		Ok((internal1, r1, internal2, r2))
+		Ok((n, max_r, p))
```

Comment:
Not sure if it's important but can we just remove `p` from the returned value?

`p` is already wrapped in `Arc`, the value itself is not moved or modified, so the caller should already have access to the original `p`- Why return it then?

Maybe you call `(Self, f64, Arc<Vector>)` a "node" in "exract_node" and you wanted this function to return the entire node, not the components of the node?

Reply:
Also, would you mind helping me understand why you call this `extract_node`? To me, this looks more like "build_node" or something like that, because it isn't extracting anything from- instead it is "build"ing a node object from the center(=p), the entries=vectors, and the distances across entries?

Reply:
I have to explain a pattern I am often using. I try to avoid concurrent operations when possible when it is possible to just move them. Let me show an example of what I am doing here:

Let me show in detail what I am achieving.

Here are the concrete two options. The question is which "caller" approach will be faster.

Option 1:

``` rust
fn caller() {
    let value = Arc::new(vec![1,,2,3,4]);
    do_something(value.clone());
    // then do something else with value
    ...
}

fn do_something(value: Arc<Vec<i32>>)  {
    // do something with the value
    ...
}
```

Option2:

``` rust
fn caller() {
     let value = Arc::new(vec![1,,2,3,4]);
     let value = do_something(value);
    // then do something else with value
    ...
}

fn do_something(value: Arc<Vec<i32>>)  -> Arc<Vec<i32>> {
    // do something with the value
    ...
    value // move the value back
}
```

Let's analyze the two options:

Option 1:

- In caller, before invoking `do_something`, the `Arc<Vec<i32>>` is cloned, leading to an atomic increment of the reference count.
- Inside `do_something`, the value is used, and upon exiting the function, the `Arc` is dropped, leading to an atomic decrement of the reference count.

Option 2:

- In caller, the Arc<Vec<i32>> is moved into do_something. No atomic operations are performed for this.
- Inside `do_something`, after using the value, it is returned. This also involves moving the `Arc` but doesn't involve any atomic operations.
- Back in `caller`, the Arc is simply reassigned to the value variable. Again, no atomic operations are involved.

Considering the above:

- Option 1 involves an atomic increment and decrement of the reference count.
- Option 2 does not involve any atomic operations on the reference count.

From a pure performance standpoint, Option 2 is expected to be faster because it avoids atomic operations on the reference count. The atomic increment and decrement operations, while efficient, still have a cost, especially when compared to simple memory moves.

Reply:
Regarding `extract_nodes`, if you look at the implementation, we are extracting the content of the node to build a new node. In this sense, you are right, it is building a node, but it is also extracting the properties of the object from the node. A perfect name would be `extract_objects_and_build_node`, but I find it a bit too big

https://github.com/surrealdb/surrealdb/blob/139024b12816cca2713ddad65998dd6172ffb56e/lib/src/idx/trees/mtree.rs#L1278-L1296

Reply:
I have implemented a micro-benchmark comparing the two approaches: [move_vs_clone](https://github.com/surrealdb/surrealdb/blob/19bc8d57f8d3d3a6e43eced2e48366baed565805/lib/benches/move_vs_clone.rs
)

According to the benchmark result, `Move` is 2x faster than `Clone` (in the benchmark).

```
cargo bench --package surrealdb --no-default-features --features kv-mem,scripting,http move_vs_clone

     Running benches/move_vs_clone.rs (target/release/deps/move_vs_clone-573638115345fc06)
move_vs_clone/move      time:   [1.8582 ms 1.8703 ms 1.8886 ms]
                        thrpt:  [529.49  elem/s 534.68  elem/s 538.15  elem/s]
move_vs_clone/clone     time:   [3.5891 ms 3.6341 ms 3.7029 ms]
                        thrpt:  [270.06  elem/s 275.17  elem/s 278.62  elem/s]

```

Reply:
@emmanuel-keller Awesome! Thanks for the clarification- it makes much more sense now! Super excited to see how much thought was put into the optimization 🌟 

I think I now understand how to use Arcs efficiently here. But still unsure if this is the right place to use Arc!

That is- I'm wondering if we can use the `.as_slice` of the vec instead of moving the entire Arc-wrapped vec here?

My rust-beginner brain is still saying you'd need to use Arc only when you need shared ownership across threads, and this function doesn't involve any multi-threading thing, no? 🤔 

Am I missing anything, or maybe is there any great pattern you found that Arc<Vec> is handy/efficient/etc in a context other than shared ownership?

Reply:
> I think I now understand how to use Arcs efficiently here. But still unsure if this is the right place to use Arc!

@mumoshu You're right, at this moment my requirement is multiple ownership but not thread safety. However, because we are running in an async context, and because I am using cache structures that use the Arc<Vector> as a key, the cached objects has to be `Sent`, which then requires the use of Arc.

Here is what happens if I replace Arc with Rc.
```
error: future cannot be sent between threads safely
   --> lib/src/api/engine/local/native.rs:96:15
    |
96  |       tokio::spawn(async move {
    |  __________________^
97  | |         let configured_root = match address.config.auth {
98  | |             Level::Root => Some(Root {
99  | |                 username: &address.config.username,
...   |
160 | |         let _ = maintenance_tx.into_send_async(()).await;
161 | |     });
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `std::marker::Send` is not implemented for `Rc<Vec<sql::number::Number>>`
note: future is not `Send` as this value is used across an await
   --> lib/src/sql/statements/analyze.rs:63:31
    |
61  |                         let mt =
    |                             -- has type `MTreeIndex` which is not `Send`
62  |                             MTreeIndex::new(&mut run, ikb, p, TreeStoreType::Traversal).await?;
63  |                         mt.statistics(&mut run).await?.into()
    |                                                 ^^^^^ await occurs here, with `mt` maybe used later
64  |                     }
    |                     - `mt` is later dropped here
note: required by a bound in `tokio::spawn`
   --> /Users/ekeller/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/task/spawn.rs:166:21
    |
164 |     pub fn spawn<T>(future: T) -> JoinHandle<T::Output>
    |            ----- required by a bound in this function
165 |     where
166 |         T: Future + Send + 'static,
    |                     ^^^^ required by this bound in `spawn`
```

> That is- I'm wondering if we can use the .as_slice of the vec instead of moving the entire Arc-wrapped vec here?

I can't because I am also using this passed object (parameter `p`) as a key in a map, which requires an Arc<Vector>

https://github.com/surrealdb/surrealdb/blob/139024b12816cca2713ddad65998dd6172ffb56e/lib/src/idx/trees/mtree.rs#L1322